### DISCUSSION TITLE
Is there a way to disable the mouse click on a button?

### DISCUSSION THREAD

**Discussion by Anindya088** [zauberzeug/nicegui#1055] [Q&A] [answered]
### Question

Hi All,

I am trying to create a login card, with inputs for username and password. For the login button, I want to disable it if the user has left any of the fields blank. 

So what I did was to use the ```on('mouseover')``` method, and then use the ```disabled``` props so that when I hover the mouse on the button, it checks for the 2 inputs and make it look like disabled. But then I realized that actually you can still click the button, that ```disabled``` is basically just changing the mouse icon (😄I did not know this before)

So now I am using a separate function just to ignore that case where both fields are blank, and combine that with my ```on_click``` method.

I am sharing my code below here, but I wanted to know if there is a proper way of disabling the mouse click in such cases. So basically if any of the fields are blank, then even if the user clicks the button then nothing happens. 

This is what i am using for the inputs and button
```python
username = ui.input(placeholder="Username").props('clearable')
password = ui.input(placeholder='Password', password=True, password_toggle_button=True).props('clearable')
err_msg = ui.label('').classes('text-xs text-red')   # for the purpose of showing error messages, blank in the beginning

signup = ui.button("Sign up",icon='login',on_click=lambda: auth_check()).classes('rounded-lg w-24 w-28')
signup.on('mouseover',lambda: input_check())
```

And then, these are the 2 functions:

```python
def input_check():
    if username.value == "" or email.value == "" or password.value == "" or pwd_repeat.value == "":
        signup.classes('disabled')
        
def auth_check():
    if username.value == "" or email.value == "" or password.value == "" or pwd_repeat.value == "":
        pass     # using this to disable mouse click
    # other checks here like username, password mismatch etc.
```     
        
Please let me know if there is a better or more proper way of doing this, which I might be missing.

Thanks,
Anindya

**Comment by falkoschindler** [comment]
Hi @Anindya088!

There are multiple steps to solving your problem:

1. There is an unfortunate confusion between the HTML attribute "disabled" and the Quasar prop "disable". Using "disabled" for Quasar elements doesn't always yield the expected result. But "disable" should work well.
2. Both, "disabled" and "disable", are _props_ (or _attributes_, respectively), but not classes. Therefore you need to set them using the `.props()` method.
3. Your checks don't enable the element again. Using an `else` clause with something like `.props(remove='disable')` should do the job.
4. It is a bit weird to do the check on mouseover. I would rather use the input's `on_change` event handler.
5. I would simply check `if not username.value` instead of `if username.value == ""`. The latter doesn't cover the initial state of `None`, while `if not username.value` covers every falsy value, i.e. `""` _and_ `None`.
6. Because of the confusion with "disable" and "disabled" we added methods and properties for enabling and disabling NiceGUI elements. So you can simply call `.enable()` or `.disable()` on elements like `ui.button`.
7. Last but not least, you can simply bind the `enabled` property to the value of the input element. This way the button is automatically enabled/disabled depending on the input value. Note that this automatically treats the input value as a boolean, so `""` and `None` will disable the button:
    ```py
    username = ui.input()
    ui.button('Submit', on_click=lambda: ui.notify('Done.')).bind_enabled_from(username, 'value')
    ```

To summarize: `.bind_enabled_from(username, 'value')` is all you need. 😊

**Comment by Anindya088** [comment]
Hi @falkoschindler ,

Wow, all of that I was trying, you can just simply put it inside one method ```.bind_enabled_from()``` ..... this is so neat, amazing !!
And thanks for the detailed steps, they are very helpful, I am learning the inner workings a bit better now 😄 

This got me thinking, and I have 2 questions where I need your advice please:
1. How to replicate this for multiple inputs? I assume I can create a function to check for True/False and pass using lambda?
2. How to update the button from disabled to enabled once the user inputs something? 

So for ex: I tried the following (I tried the class/object style because I could not figure out how to use the lambda function here)

```python
username = ui.input()
password = ui.input()

class Val:
    def __init__(self):
        fields = [username, password]
        show_submit = any(field.value == '' for field in fields)
        self.filled = not show_submit

form_val = Val()
submit = ui.button('Submit', on_click=lambda: ui.notify(form_val.filled)).bind_enabled_from(form_val,'filled')
```
So what I tried here, is to check for both input boxes, then pass that logic to an object ```form_val``` and then use it to bind the visibility. This works in the beginning when I load the page, but then when I type in something, the submit button is still disabled.

I think this is because, after I type in something, my code is not checking the condition again anymore. Is that the issue? 
Can you please advise on how to resolve this? And also, probably instead of using class/object make it simpler and use a lamda function inside ```bind_visibility_from()```?

Thanks,
Anindya

**Reply by falkoschindler** [reply, marked as answer]
> I think this is because, after I type in something, my code is not checking the condition again anymore.

Yes, you're right. The attribute `self.filled` is only set once.

Unfortunately, there is currently no direct support for binding from multiple sources. But your approach is a reasonable workaround. We just need to make `filled` a property that is re-evaluated every time it is read:
```py
username = ui.input()
password = ui.input(password=True)

class Val:
    @property
    def filled(self):
        fields = [username, password]
        self.filled = all(field.value for field in fields)

form_val = Val()
ui.button('Submit', on_click=lambda: ui.notify('Done.')).bind_enabled_from(form_val, 'filled')
```

A less advanced alternative might be to simply use a change handler that updates the button:
```py
def update_button():
    submit.enabled = username.value and password.value

username = ui.input(on_change=update_button)
password = ui.input(on_change=update_button, password=True)
submit = ui.button('Submit', on_click=lambda: ui.notify('Done.'))
update_button()
```

Similarly, you can make use of the rather new `ui.refreshable` decorator, even though it is not a huge advantage in this specific scenario:
```py
@ui.refreshable
def submit_button():
    button = ui.button('Submit', on_click=lambda: ui.notify('Done.'))
    button.enabled = username.value and password.value

username = ui.input(on_change=submit_button.refresh)
password = ui.input(on_change=submit_button.refresh, password=True)
submit_button()
```

**Reply by Anindya088** [reply]
Hi @falkoschindler , thanks so much for the explanations. The 2nd option that you gave, is so much easier 😄 I think I will just use that. 
Thanks again for the help.

Regards,
Anindya

**Reply by falkoschindler** [reply]
You're welcome, @Anindya088! Yes, I would prefer 2nd option as well. 🙂
